pkgs/applications/editors/sublime/4/common.nix

```diff
@@ -0,0 +1,148 @@
+, writeShellScript, common-updater-scripts, curl
+, openssl, bzip2, bash, unzip, zip
+}:
+
+let
+  pname = "sublimetext4";
+  packageAttribute = "sublime4${lib.optionalString dev "-dev"}";
+  binaries = [ "sublime_text" "plugin_host-3.3" "plugin_host-3.8" "crash_reporter" ];
+  primaryBinary = "sublime_text";
+  primaryBinaryAliases = [ "subl" "sublime" "sublime4" ];
+  downloadUrl = "https://download.sublimetext.com/sublime_text_build_${buildVersion}_${arch}.tar.xz";
+  versionUrl = "https://download.sublimetext.com/latest/${if dev then "dev" else "stable"}";
+  versionFile = builtins.toString ./packages.nix;
+  archSha256 = {
+    "aarch64-linux" = aarch64sha256;
+    "x86_64-linux" = x64sha256;
+  }.${stdenv.hostPlatform.system};
+  arch = {
+    "aarch64-linux" = "arm64";
+    "x86_64-linux" = "x64";
+  }.${stdenv.hostPlatform.system};
+
+  libPath = lib.makeLibraryPath [ xorg.libX11 xorg.libXtst glib libglvnd openssl gtk3 cairo pango ];
+in let
+  binaryPackage = stdenv.mkDerivation {
+    pname = "${pname}-bin";
+    version = buildVersion;
+
+    src = fetchurl {
+      url = downloadUrl;
+      sha256 = archSha256;
+    };
+
+    dontStrip = true;
+    dontPatchELF = true;
+    buildInputs = [ glib gtk3 ]; # for GSETTINGS_SCHEMAS_PATH
+    nativeBuildInputs = [ zip unzip makeWrapper wrapGAppsHook ];
+
+    # make exec.py in Default.sublime-package use own bash with an LD_PRELOAD instead of "/bin/bash"
+    patchPhase = ''
+      runHook prePatch
+
+      # TODO: Should not be necessary even in 3
+      mkdir Default.sublime-package-fix
+      ( cd Default.sublime-package-fix
+        unzip -q ../Packages/Default.sublime-package
+        substituteInPlace "exec.py" --replace \
+          "[\"/bin/bash\"" \
+          "[\"$out/sublime_bash\""
+        zip -q ../Packages/Default.sublime-package **/*
+      )
+      rm -r Default.sublime-package-fix
+
+      runHook postPatch
+    '';
+
+    buildPhase = ''
+      runHook preBuild
+
+      for binary in ${ builtins.concatStringsSep " " binaries }; do
+        patchelf \
+          --interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
+          --set-rpath ${libPath}:${stdenv.cc.cc.lib}/lib${lib.optionalString stdenv.is64bit "64"} \
+          $binary
+      done
+
+      # Rewrite pkexec argument. Note that we cannot delete bytes in binary.
+      sed -i -e 's,/bin/cp\x00,cp\x00\x00\x00\x00\x00\x00,g' ${primaryBinary}
+
+      runHook postBuild
+    '';
+
+    installPhase = ''
+      runHook preInstall
+
+      mkdir -p $out
+      cp -r * $out/
+
+      # We can't just call /usr/bin/env bash because a relocation error occurs
+      # when trying to run a build from within Sublime Text
+      ln -s ${bash}/bin/bash $out/sublime_bash
+
+      runHook postInstall
+    '';
+
+    dontWrapGApps = true; # non-standard location, need to wrap the executables manually
+
+    postFixup = ''
+      sed -i 's#/usr/bin/pkexec#pkexec\x00\x00\x00\x00\x00\x00\x00\x00\x00#g' "$out/${primaryBinary}"
```

Comment:
This no longer works; `strace` shows that it `stat()`s the file first, which fails:

```c
stat("/usr/bin/pkexec", 0x7f3959ff98b0) = -1 ENOENT (No such file or directory)
```

We may need to look again into `libredirect`, undoing https://github.com/NixOS/nixpkgs/pull/119351#discussion_r612799394, because NixOS's path to `pkexec` is longer than this one.

Reply:
(This is the same problem as https://github.com/NixOS/nixpkgs/pull/10517#issuecomment-150702971 by @demin-dmitriy)

Reply:
When I try to put `libredirect` back in place, it doesn't work, seems to have no effect; sublime is still using the unredirected path. Same problem for current `sublime3`.

Anybody know why that is?

Reply:
There are two main problems with libredirect:

- it is incomplete, we need to implement more functions like https://github.com/NixOS/nixpkgs/commit/90f3a237eb51d7e3173f1502d468e99ff777d1b9
- it is inherited by children and symbol clashing occurs – this happens a lot for me with ST (crashing child `git`)

Reply:
If it still tries to use `/usr/bin/pkexec`, the sed call did not work for some reason.

Reply:
> If it still tries to use `/usr/bin/pkexec`, the sed call did not work for some reason.

@jtojnar That's not what I mean, I meant to say:

* The `sed` call itself works, but it does not make `pkexec` work, because Sublime `stat()`s it before it opens it. And `stat("pkexec")` returns `No such file or directory` because a file `pkexec` does not exist in whatever directory the user started Sublime from.
* If I _remove_ the `sed` call and bring back the `libredirect` approach that replaces redirects `/usr/bin/pkexec` by a NixOS specific path at runtime, then it still does `stat("/usr/bin/pkexec")` instead of using the redirected path, even though [a `stat()` redirect exists in `libredirect.c`](https://github.com/NixOS/nixpkgs/blob/dfc658fd2e6eee444c2bd7258c3770613b16f3fd/pkgs/build-support/libredirect/libredirect.c#L152).

Reply:
I used gdb to see how the syscall is made, based on [this](https://stackoverflow.com/questions/37743596/gdb-catch-syscall-condition-and-string-comparisson), which shows how to break on syscall string argument values (the file path given to `stat()`):

```
$ sudo gdb -p $(pgrep sublime_text)
(gdb) catch syscall stat
(gdb) condition 1 $_streq((char *)$rdi, "/usr/bin/pkexec")
(gdb) continue

# pressing save now in my non-root user's Sublime on /etc/nixos/configuration.nix, which is a root-owned file

Thread 14 ".sublime_text-w" hit Catchpoint 1 (call to syscall stat), 0x00007f6d44152555 in __xstat64 ()
   from /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/libc.so.6

(gdb) backtrace
#0  0x00007f6d44152555 in __xstat64 ()
   from /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/libc.so.6
#1  0x00007f6d44702df1 in __xstat64 ()
   from /nix/store/f48az5jm3c6jlnzgcg526xci8y66k51m-libredirect-0/lib/libredirect.so
#2  0x000000000063f37a in ?? ()
#3  0x0000000000702c13 in ?? ()
#4  0x000000000081ee7d in ?? ()
#5  0x000000000086e0c4 in ?? ()
#6  0x00000000004fd49f in ?? ()
#7  0x00007f6d4422fe9e in start_thread ()
   from /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/libpthread.so.0
#8  0x00007f6d4416149f in clone ()
   from /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/libc.so.6
```

This confirmed that indeed `__xstat64` is the one to debug.

Reply:
I found the reason it doesn't work:

https://github.com/NixOS/nixpkgs/blob/dfc658fd2e6eee444c2bd7258c3770613b16f3fd/pkgs/build-support/libredirect/libredirect.c#L27

`unsetenv("NIX_REDIRECTS");` makes that the redirect is turned off for spawned subprocesses.

That seems wrong.

Reply:
PR to fix that: #124157

Reply:
> it is inherited by children and symbol clashing occurs – this happens a lot for me with ST (crashing child `git`)

@jtojnar Could you check if these problems still occur for you? I made PR #124160 to try it easily.

Reply:
I also found that Sublime on startup reads `/proc/self/exe` and does `posix_spawnp` on itself.

This is probably why the `unsetenv("NIX_REDIRECTS")` broke it.

Reply:
@jtojnar Another strategy might be to optionally disable the `LD_PRELOAD` for subprocesses, so that it is e.g. only active for the `sublime_text` process but not for started `git` or `gksudo` processes (`gksudo` also segfaults for me when `LD_PRELOAD` is active).

For example, this modification to `libredirect.c` worked for me to make the sublime+gksudo workflow work:

```c
int posix_spawnp(pid_t * pid, const char * file,
// ...
    if (!string_endswith(file, ".sublime_text-wrapped")) {
        unsetenv("LD_PRELOAD");
    }
```

We could add an environment variable that controls wihch binaries (like `.sublime_text-wrapped`) have this active, or generate an application-specific `libredirect.c` to insert such custom logic.

What do you think?

Reply:
That sounds like a good compromise.

Reply:
I pushed some experiments into this branch:

https://github.com/nh2/nixpkgs/tree/sublime4-reintroduce-libredirect-wip